src/main/java/com/marklogic/client/ext/datamovement/listener/ProgressQueryBatchListener.java

```diff
@@ -0,0 +1,154 @@
+package com.marklogic.client.ext.datamovement.listener;
+
+import com.marklogic.client.datamovement.QueryBatch;
+import com.marklogic.client.datamovement.QueryBatchListener;
+import com.marklogic.client.datamovement.QueryBatcher;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+
+/**
+ * QueryBatchListener for receiving progress updates as QueryBatch instances are processed. Progress updates are
+ * communicated via instances of the nested static ProgressUpdate class, which are given to an implementation of the
+ * nested static ProgressUpdateListener callback interface.
+ */
+public class ProgressQueryBatchListener implements QueryBatchListener {
+
+	private ProgressUpdateListener progressListener;
+
+	private long startTime;
+	private long totalResults;
+	private long resultsCount;
+
+	/**
+	 * Use this constructor for when the total number of results isn't known ahead of time.
+	 *
+	 * @param progressListener
+	 */
+	public ProgressQueryBatchListener(ProgressUpdateListener progressListener) {
+		this(progressListener, 0);
+	}
+
+	/**
+	 * Use this constructor for when the total number of results is known ahead of time. The ProgressUpdate objects that
+	 * are emitted will have more interesting information as a result.
+	 *
+	 * @param progressListener
+	 * @param totalResults
+	 */
+	public ProgressQueryBatchListener(ProgressUpdateListener progressListener, long totalResults) {
+		this.progressListener = progressListener;
+		this.totalResults = totalResults;
+	}
+
+	/**
+	 * Initializes the start time so that each ProgressUpdate knows how long it occurred after the job was started.
+	 *
+	 * @param queryBatcher
+	 */
+	@Override
+	public void initializeListener(QueryBatcher queryBatcher) {
+		startTime = System.currentTimeMillis();
+	}
+
+	/**
+	 * Batches arrive in random order, so a ProgressUpdate is created and sent to the ProgressUpdateListener only if the
+	 * value of "getJobResultsSoFar" on the QueryBatch exceeds the number of results seen so far.
+	 * <p>
+	 * For example, if there are 2 batches, and batch 2 is processed first by this listener followed by batch 1, a
+	 * ProgressUpdate is only created when batch 2 is processed.
+	 *
+	 * @param batch
+	 */
+	@Override
+	public void processEvent(QueryBatch batch) {
+		boolean resultsCountUpdated = updateResultsCountIfNecessary(batch);
+		if (resultsCountUpdated && progressListener != null) {
+			double timeSoFar = ((double) System.currentTimeMillis() - startTime) / 1000;
+			progressListener.onUpdatedProgress(new ProgressUpdate(startTime, totalResults, batch.getJobResultsSoFar(), timeSoFar));
+		}
+	}
+
+	/**
+	 * This is protected so that a subclass can make this synchronized if desired. Without this being synchronized, there
+	 * is a slight chance of a race condition between two threads where the results count could be set to an incorrect
```

Comment:
Can we use AtomicLong and use compareAndSet method? I'm just wondering if that will solve the race condition. If it is not too big of a problem, then we can ignore.

Reply:
I'll check that out